src/bazel/bazel_quickpick.ts

```diff
@@ -169,10 +173,108 @@ export async function queryQuickPickPackage({
   const packagePaths = await new BazelQuery(
     getDefaultBazelExecutablePath(),
     workspaceInfo.workspaceFolder.uri.fsPath,
-  ).queryPackages(query ?? "//...");
+  ).queryPackages(query ?? "//...", { abortSignal });
 
   // Sort the labels so the QuickPick is ordered.
   return packagePaths
     .sort()
     .map((target) => new BazelTargetQuickPick("//" + target, workspaceInfo));
 }
+
+/**
+ * Shows a QuickPick of Bazel labels that dynamically updates its items based on the user's input.
+ *
+ * @param options Configuration options for the QuickPick
+ * @returns A promise that resolves with the selected BazelTargetQuickPick, or undefined if no selection was made
+ */
+export function showDynamicQuickPick(options: {
+  query?: string;
+  queryFunctor: (params: QuickPickParams) => Promise<BazelTargetQuickPick[]>;
+  workspaceInfo?: BazelWorkspaceInfo;
+}): Promise<BazelTargetQuickPick | undefined> {
+  const quickPick = vscode.window.createQuickPick<BazelTargetQuickPick>();
+  quickPick.title = "Select a Bazel target";
+  quickPick.placeholder = "Start typing to search for targets...";
+  quickPick.matchOnDescription = true;
+  quickPick.matchOnDetail = true;
+
+  let abortController: AbortController | undefined = new AbortController();
+  let timeout: NodeJS.Timeout | undefined;
+
+  const updateQuickOptions = async (value: string): Promise<void> => {
+    // Cancel any previous query
+    abortController?.abort();
+
+    // Store the current abort controller to ensure we don't cancel a new query
+    const currentAbortController = new AbortController();
+    abortController = currentAbortController;
+
+    // Show loading state
+    quickPick.busy = true;
+    quickPick.items = [];
+
+    // Process the input: keep everything before the last slash and add '...'
+    const newQuery = value.includes("/")
+      ? options.query.replace(
```

Comment:
Looks like we sometimes don't pass a query here but don't account for this by defaulting `options.query`. Note we don't have strict null checks enabled so this won't be caught by typescript unfortunately!